Onboard Windows PR tests to Azure Pipelines YAML and Scale Sets

BillyONeal opened the pull request:
**Describe the pull request**

This PR contains supporting infrastructure to use Azure Virtual Machine Scale Sets for vcpkg's PR builds, which should both improve our PR build times and reduce Azure spending by shutting down machines when they aren't being used.

Included is a script that sets up all vcpkg's Azure infrastructure for Windows PR tests, and several updates to baselines. The baseline updates are generally caused by an updated copy of the MSVC++ compiler caused by updating the VMs, but some are caused by missed failures only detected now because this did a cleared out archives directory first.

Some of the build infrastructure isn't what I'd call 'pretty' (e.g. we're split into more scripts and such than I'd like) but this mirrors how our existing PR system works.

It is expected that the existing vcpkg Windows PR system will hate these baseline updates so we'll need to merge this, then remove that (duplicate) workflow immediately afterwards, then delete all the Windows VMs powering the old infrastructure.

- Which triplets are supported/not supported?

All the Windows ones so far.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

cbezault commented on scripts/azure-pipelines/analyze-test-results.ps1:
```diff
@@ -0,0 +1,661 @@
+
+    [xml]$xmlContents = Get-Content $xmlFilename
+
+    # This currently only supports one collection per assempbly, which is the way
```
Typo

cbezault commented on scripts/azure-pipelines/analyze-test-results.ps1:
```diff
@@ -0,0 +1,661 @@
+        $name = ($test.name -replace ":.*$")
+
+        # Reconstruct the original BuildResult enumeration (defined in Build.h)
+        #   failure.message - why test failed (valid only on test failure)
+        #   reason - why test skipped (valid only when test is skipped)
```
Grammar on these two lines.

cbezault commented on scripts/azure-pipelines/analyze-test-results.ps1:
```diff
@@ -0,0 +1,661 @@
+        $abi_tag = ""
+        $features = ""
+        foreach ( $trait in $test.traits.trait) {
+            #Write-Verbose "$($trait.name):$($trait.value)"
```
Commented out code? If it's worth keeping it for debugging there's a standard way to guard debug code in ADO pipelines using the `System.Debug` variable. (I think it gets injected as an environment variable)

cbezault commented on scripts/azure-pipelines/analyze-test-results.ps1:
```diff
@@ -0,0 +1,661 @@
+    }
+}
+
+# Creates an object that represents the baseline expectations
```
What about ignore?

BillyONeal replied:
No idea. This comment is unedited from the form in vcpkg-utils :/

cbezault replied:
No worries.

cbezault commented on scripts/azure-pipelines/analyze-test-results.ps1:
```diff
@@ -0,0 +1,661 @@
+
+    Write-Verbose "analyzing $($currentTests.count) tests"
+
+    foreach ( $key in $currentTests.keys) { # limiting range for testing
```
Leftover comment?

cbezault commented on scripts/azure-pipelines/analyze-test-results.ps1:
```diff
@@ -0,0 +1,661 @@
+            $lookupTable = @{
+                'Pass' = @{
+                    'Fail' = @('Fail', "Test passes in baseline but fails in current run. If expected update ci.baseline.txt with '$($key):$($current.collectionName)=fail'");
+                    'Skip' = @($null, 'Test was skipped due to missing dependencies')
```
I think this line should also be a failure state.

BillyONeal replied:
I'm not changing this vs. the existing infrastructure.

cbezault replied:
After thinking about this it is fine. It's not that it's strictly not a failure but just that it isn't worth reporting as a failure in the CI for one of the following reasons.

1. You marked a dependency as explicitly skipped (so of course this package failed to build).
2. Some package that was meant to be a PASS failed to build and so this package could not build (then the CI fails because the dependency will failed to build).

It might be a nice improvement to have a way of insisting a package succeed building in the checked-in baseline. The way things are now if a package is marked as `ignore` and it fails to build but one of the packages that depend on it is marked as `pass` but doesn't get built because it's dependency failed the whole CI still passes.

cbezault commented on scripts/azure-pipelines/windows/azure-pipelines.yml:
```diff
@@ -0,0 +1,149 @@
+jobs:
+- job: ${{ parameters.jobName }}
+  pool:
+    name: PrWin-2020-04-13
```
Come up with a better name for the pool unless it's meant to be the same as the one in `azure-pipelines.yml`. If so use `$(agentPool).

BillyONeal replied:
I ran into name length limits vs. what I wanted to name the pool (something like vcpkg-windows-prtest-date). Do you have suggestions on what you want it named instead?

cbezault replied:
The name is fine as long as it meant to be the same pool as the one defined in the `azure-pipelines.yml` (right now that one is `PrWin-2020-04-05-1`).

BillyONeal replied:
Oh, derp. Yeah that one should go.

cbezault commented on scripts/azure-pipelines/windows/azure-pipelines.yml:
```diff
@@ -0,0 +1,149 @@
+
+  steps:
+  - powershell: |
+      $StorageAccountName= $env:StorageAccountName
```
```suggestion
      $StorageAccountName = $env:StorageAccountName
```

cbezault commented on scripts/azure-pipelines/windows/ci-step.ps1:
```diff
@@ -0,0 +1,104 @@
+        # WORKAROUND: the x86-windows flavors of these are needed for all
+        # cross-compilation, but they are not auto-installed.
+        # Install them so the CI succeeds
+        ./vcpkg install "protobuf:x86-windows" "boost-build:x86-windows" "sqlite3:x86-windows"
```
These are only needed in the `x86` or `x64` cross compilation to `arm`, `arm64` or `uwp` scenarios. Might be worth updating the comment or making them only get installed when necessary.

Unless of course it is dependent ports themselves artificially adding the `x86` requirement then that should also be noted.

BillyONeal replied:
I don't really know how to edit the comment as you describe.

cbezault replied:
Meh whatever.

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+    $ExitCode
+  )
+
+  if ($ExitCode -eq 0 -or $ExitCode -eq 3010) {
```
What's an `ExitCode` of `3010` mean?

BillyONeal replied:
I don't remember exactly but

*plug plug plug decoder ring*

https://james.darpinian.com/decoder/?q=3010

I think `ERROR_SUCCESS_REBOOT_REQUIRED` ?

strega-nil replied:
ah, cool. Could you put that in a comment?

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+  )
+
+  try {
+    Write-Output 'Downloading Visual Studio...'
```
All of these `Write-Output`s should be `Out-Host` -- `Output` is passed to the next thing in the pipeline, whereas this is informational, and thus should be printed to the screen.

BillyONeal replied:
No, they explicitly must not be xxxHost, because in some environments there is no host. For example emitting Write-Host during an Azure `Invoke-VmCommand` fails.

BillyONeal replied:
(And also the performance of Xxx-Host is substantially worse than Xxx-Output, see http://www.jsnover.com/blog/2013/12/07/write-host-considered-harmful/ for additional details)

strega-nil replied:
This is not true past Powershell 5.0.0; `Write-Host` is a different way of calling `Write-Information`, and `Out-Host` has never acted like `Write-Host` used to, aiui

BillyONeal replied:
Then maybe Azure doesn't run that version. All I know is if you put Write-Host in a script passed to `Invoke-AzVmCommand` (as this one is), the script silently crashes there.

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+    }
+
+    $proc = Start-Process -FilePath cmd.exe -ArgumentList $args -Wait -PassThru
+    PrintMsiExitCodeMessage $proc.ExitCode
```
Is there a reason you aren't doing something along the lines of 

```ps1
cmd.exe @args
PrintMsiExitCodeMessage $LastExitCode
```
? I think that they should be equivalent.

BillyONeal replied:
I believe these workarounds were to deal with the script getting stuck due to installers getting confused by being attached to a script console.

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+    PrintMsiExitCodeMessage $proc.ExitCode
+  }
+  catch {
+    Write-Output 'Failed to install Visual Studio!'
```
This should be:

```ps1
Write-Error "Failed to install Visual Studio!
$($_.Exception.Message)"
```

BillyONeal replied:
I applied the Write-Error change but it is likely that this won't work if a failure is encountered. Any cmdlets that interact with the host (like Write-Host) silently crash the script in the environment in which this script runs, so if Write-Error does that it might doom any feedback here.

(I applied the change anyway because even in the best of times the feedback we get from this is not amazing; Invoke-AzVmCommand likes to drop the output :( )

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+# SPDX-License-Identifier: MIT
+#
+# Sets up a machine in preparation to become a build machine image, optionally switching to
+# AdminUser first.
```
Probably want to add `[CmdletBinding()]` to make this far nicer to call from powershell.

strega-nil replied:
Also, this would preferably have a doc comment.

BillyONeal replied:
I'm unfortunately unfamilar with doc comments in PS land.

strega-nil replied:
I'll link you some examples in teams.

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+  [string]$AdminUserPassword = $null,
+  [string]$StorageAccountName = $null,
+  [string]$StorageAccountKey = $null
+)
```
May want to set `$ErrorActionPreference = 'Stop'` here.

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+    Write-Output "Installing $Name..."
+    $args = @('/i', $msiPath, '/norestart', '/quiet', '/qn')
+    $proc = Start-Process -FilePath 'msiexec.exe' -ArgumentList $args -Wait -PassThru
+    PrintMsiExitCodeMessage $proc.ExitCode
```
The same comment as above with `$LastExitCode`

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+  }
+  catch {
+    Write-Output "Failed to install $Name!"
+    Write-Output $_.Exception.Message
```
The same comment as above with `Write-Error`.

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,320 @@
+  }
+}
+
+Function InstallCuda {
```
This is a lot of copy-paste redundancy; is there any way we could merge these into one?

BillyONeal replied:
If you have suggestions that would meaningfully improve readability here I'm all for merging them but they are just subtly different enough that I didn't see obvious means of deduplication. It's certainly possible that someone more familiar with PS would have better ideas.

Keep in mind testing this thing is a nightmare because it takes at least an hour per attempt (and 12 hours per attempt if one actually wants to ensure the right bits were installed) so I'm reluctant to move things around merely for the sake of being terse.

strega-nil replied:
Ah, fair enough.

strega-nil commented on scripts/azure-pipelines/windows/create-vmss.ps1:
```diff
@@ -0,0 +1,396 @@
+$CurrentProgress = 1
+
+function Find-ResourceGroupNameCollision {
+  Param([string]$Test, $Resources)
```
This should probably be `[String[]]$Resources`?

strega-nil commented on scripts/azure-pipelines/windows/create-vmss.ps1:
```diff
@@ -0,0 +1,396 @@
+$TotalProgress = 12
+$CurrentProgress = 1
+
+function Find-ResourceGroupNameCollision {
```
this function should probably be called `Test-ResourceGroupNameUnique`?

BillyONeal replied:
This returns true if there is a collision, with that name we would want it to return the opposite ?

strega-nil replied:
Makes sense; `Test-ResourceGroupNameCollision` works too.

strega-nil commented on scripts/azure-pipelines/windows/create-vmss.ps1:
```diff
@@ -0,0 +1,396 @@
+  $Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+  $result = ''
+  for ($idx = 0; $idx -lt $Length; $idx++) {
+    $result += $Chars[(Get-Random -Minimum 0 -Maximum $Chars.Length)]
```
`Get-Random` is not cryptographically secure, and should probably not be used for generating passwords. Instead, use [RNGCryptoServiceProvider].

[RNGCryptoServiceProvider]: https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rngcryptoserviceprovider?redirectedfrom=MSDN&view=netframework-4.8

BillyONeal replied:
RNGCryptoServiceProvider is not available in Powershell Core or Cloud Shell.

I agree this is suboptimal but it's still better than prompting the user for a password.

strega-nil commented on scripts/azure-pipelines/windows/create-vmss.ps1:
```diff
@@ -0,0 +1,396 @@
+  return $result
+}
+
+function Start-WaitForShutdown {
```
Rename this `Wait-Shutdown` I think.

strega-nil commented on scripts/azure-pipelines/windows/create-vmss.ps1:
```diff
@@ -0,0 +1,396 @@
+  $RawName = $RawName.ToLowerInvariant()
+
+  $result = ''
+  foreach ($ch in $RawName.ToCharArray()) {
```
`$RawName.replace("-", "")`

strega-nil commented on scripts/azure-pipelines/windows/create-vmss.ps1:
```diff
@@ -0,0 +1,396 @@
+$Credential = New-Object System.Management.Automation.PSCredential ("AdminUser", $AdminPWSecure)
+
+####################################################################################################
+Write-Progress `
```
Good use of `Write-Progress` :)

strega-nil commented on scripts/azure-pipelines/windows/ci-step.ps1:
```diff
@@ -0,0 +1,104 @@
+[CmdletBinding()]
+param(
+    [Parameter(Mandatory = $true)][string]$Triplet,
+    [AllowEmptyString()][string]$OnlyIncludePorts = "",
```
Since these aren't mandatory, they don't need to `AllowEmptyString`.

strega-nil commented on scripts/azure-pipelines/generate-skip-list.ps1:
```diff
@@ -0,0 +1,55 @@
+#
+
+Param(
+    [Parameter(Mandatory = $false)][string]$Triplet,
```
You don't need `Mandatory = $false`; they're non-mandatory by default.

strega-nil commented on scripts/azure-pipelines/generate-skip-list.ps1:
```diff
@@ -0,0 +1,55 @@
+)
+
+if (-not (Test-Path -Path $BaselineFile)) {
+    Write-Error "Unable to find baseline file $BaselineFile"
```
IMO, this should stop the CI in its tracks; we should set `ErrorActionPreference` to `Stop` or pass it explicitly here.

strega-nil reviewed (changes requested):
Really good generally, but the powershell needs some work; I'm not sure if it's yours or you ported it from the old system. If you'd like, I can make the changes I suggest to update it to more idiomatic powershell.

Thanks Billy, this PR is dope and definitely needed. :heart:

BillyONeal commented:
@strega-nil The scripts `create-vmss.ps1`, `provision-image.ps1`, and `sysprep.ps1` are new (effectively copied from microsoft/STL equivalents with minor changes to support vcpkg rather than the STL). The other scripts are almost unmodified from the old vcpkg-utils repo. For those 3 new ones I'm happy to fix tons of nitpicks, for the others I would prefer to avoid blocking this on such changes as they were the status quo if possible.

cbezault commented:
Neck yes green green green

Neumann-A commented on scripts/ci.baseline.txt:
```diff
@@ -1467,6 +1483,32 @@ qpid-proton:x64-windows-static=fail
 qscintilla:x64-linux=fail
 qt5-activeqt:x64-linux=fail
 qt5-activeqt:x64-osx=fail
+#Starting package 142/820: qt5-declarative:x86-windows
+#Building package qt5-declarative[core]:x86-windows...
+#Using cached binary package: C:\agent\_work\1\s\archives\94\9428e63fede20a51ac0631a9d94d8773e593dd06.zip
+#Building package qt5-declarative[core]:x86-windows... done
+#Installing package qt5-declarative[core]:x86-windows...
+#    The following files are already installed in C:/agent/_work/1/s/installed/x86-windows and are in conflict with qt5-declarative:x86-windows
+#
+#Installed by qt5-activeqt:x86-windows
+#    tools/qt5/bin/Qt5Gui.dll
+#    tools/qt5/bin/Qt5Widgets.dll
+#    tools/qt5/bin/bz2.dll
+#    tools/qt5/bin/freetype.dll
+#    tools/qt5/bin/glib-2.dll
+#    tools/qt5/bin/harfbuzz.dll
+#    tools/qt5/bin/jpeg62.dll
+#    tools/qt5/bin/libcharset.dll
+#    tools/qt5/bin/libiconv.dll
+#    tools/qt5/bin/libintl.dll
+#    tools/qt5/bin/libpng16.dll
+#    tools/qt5/bin/pcre.dll
+#    tools/qt5/bin/plugins/imageformats/qgif.dll
+#    tools/qt5/bin/plugins/imageformats/qico.dll
+#    tools/qt5/bin/plugins/imageformats/qjpeg.dll
+#    tools/qt5/bin/plugins/platforms/qwindows.dll
+#    tools/qt5/bin/plugins/styles/qwindowsvistastyle.dll
+qt5-declarative:x86-windows=skip
```
That is fishy. There is a script at the end of https://github.com/microsoft/vcpkg/blob/master/ports/qt5-base/cmake/qt_build_submodule.cmake that only dlls which are not yet installed get installed. This could mean that the binary packages of qt5 have not been build in the same go. 

Somebody should finally decide that it is ok to install tools into bin. Than all those unnecessary dll copying would go away. 

Since qt5-declarative is more important than qt5-activeqt please skip qt5-activeqt instead

BillyONeal replied:
>that only dlls which are not yet installed get installed.

That seems broken -- if the packages were built and cached at different times, the 'installed' directory might not have contained those DLLs for both of them.

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,453 @@
+
+.PARAMETER StorageAccountName
+The name of the storage account. Stored in
+HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\Environment\StorageAccountName
```
```suggestion
```

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,453 @@
+script assumes it's running on an administrator account.
+
+.PARAMETER StorageAccountName
+The name of the storage account. Stored in
```
```suggestion
The name of the storage account. Stored in the environment variable %StorageAccountName%.
```

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,453 @@
+Used by the CI system to access the global storage.
+
+.PARAMETER StorageAccountKey
+The key of the storage account. Stored in
```
```suggestion
The key of the storage account. Stored in the environment variable %StorageAccountKey%.
```

strega-nil commented on scripts/azure-pipelines/windows/provision-image.ps1:
```diff
@@ -0,0 +1,453 @@
+
+.PARAMETER StorageAccountKey
+The key of the storage account. Stored in
+HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\Environment\StorageAccountKey
```
```suggestion
```

strega-nil reviewed (commented):
These docs should be changed.

strega-nil reviewed (approved):
LGTM!

BillyONeal commented:
/azp run

BillyONeal committed:
Change supporting infrastructure to use Azure Virtual Machine Scale Sets for vcpkg's PR builds, which should both improve our PR build times and reduce Azure spending by shutting down machines when they aren't being used.

Included is a script that sets up all vcpkg's Azure infrastructure for Windows PR tests, and several updates to baselines. The baseline updates are generally caused by an updated copy of the MSVC++ compiler caused by updating the VMs, but some are caused by missed failures only detected now because this did a cleared out archives directory first.

Some of the build infrastructure isn't what I'd call 'pretty' (e.g. we're split into more scripts and such than I'd like) but this mirrors how our existing PR system works.

It is expected that the existing vcpkg Windows PR system will hate these baseline updates so we'll need to merge this, then remove that (duplicate) workflow immediately afterwards, then delete all the Windows VMs powering the old infrastructure.

BillyONeal committed:
WIP

BillyONeal committed:
WIP

BillyONeal committed:
WIP

BillyONeal committed:
WIP
